sklearn/preprocessing/_data.py

```diff
@@ -3181,6 +3200,8 @@ def _neg_log_likelihood(lmbda):
             function of lambda."""
             x_trans = self._yeo_johnson_transform(x, lmbda)
             n_samples = x.shape[0]
+            if np.ptp(x_trans) == 0:
+                return -np.inf
```

Comment:
Isn't this check too strict?
Also, shoulnt we check that x_trans is constant, instead of only equal to 0?

Reply:
> Also, shoulnt we check that x_trans is constant, instead of only equal to 0?

`np.ptp(x_trans) == 0.0` if and only if `x_trans` is constant, no?

In any case it would be worth adding a comment to explain why we test if `x_trans` is constant here (that is to avoid a low level `ZeroDivisionWarning` and instead raise a more informative `UserWarning`.

Reply:
Maybe we could compute `x_trans.var()` instead of `np.ptp(x_trans)` and reuse `x_trans_var` below if non-zero. That would make the code slightly easier to follow.

Reply:
I think that's what  #15385  did.

Reply:
I was trying to avoid the magic factor from https://github.com/scikit-learn/scikit-learn/pull/15385/files#r339817240

Maybe the more pythonic thing to do is to catch the `ZeroDivisionWarning` and then return a value. "Ask for forgiveness not permission"